### ISSUE TITLE
Non-existent subcommand exit code behavior changed in 1.13+ (1.13.0-rc1)

### ISSUE THREAD

**Issue by tianon** [docker/docker#28329]
On previous versions of Docker, the validity of a subcommand could be determined by invoking `docker <subcommand> --help` and checking the exit code.

For existing commands (such as `docker image --help`), the exit code was zero indicating success.

For invalid commands (such as `docker invalid --help`), the exit code was non-zero, as expected.

In v1.13.0-rc1, this is no longer the case:
```console
$ ./docker-1.12.3 images --help > /dev/null; echo $?
0
$ ./docker-1.12.3 invalid --help > /dev/null; echo $?
docker: 'invalid' is not a docker command.
See 'docker --help'.
1
$ ./docker-1.13.0-rc1 images --help > /dev/null; echo $?
0
$ ./docker-1.13.0-rc1 invalid --help > /dev/null; echo $?
0
```

Is this behavior change expected or accidental?

**Comment by cpuguy83** [comment]
ping @dnephin

**Comment by runcom** [comment]
Ping @vdemeester I remember you did a pr against 1.12.x about this

**Comment by dnephin** [comment]
I suspect #27745, cc @vieux

**Comment by vdemeester** [comment]
@dnephin I think it's more #25354 (https://github.com/docker/docker/pull/25354/files#diff-2d5500e8884566d26d571eebe5f41d4cL68) and the fact that `cobra` doesn't care at anything if `--help` is present (see https://github.com/dnephin/cobra/blob/master/command.go#L602).

**Comment by dnephin** [comment]
Ah, you're right. Cobra has this:  https://github.com/dnephin/cobra/blob/master/command.go#L746-L749 which might be masking the error response

**Comment by dnephin** [comment]
This is going to be a bit involved to fix. Is this behaviour really important, or is it just important to have some way to test for the existence of subcommands?

**Comment by tianon** [comment]
@dnephin I just use it to test for the existence of subcommands in a client binary, and this has been the only way so far, so if there were some other clean way introduced, I'm not going to be picky :innocent: :heart:

**Comment by runcom** [comment]
> This is going to be a bit involved to fix. Is this behaviour really important, or is it just important to have some way to test for the existence of subcommands?

can we just revert the original PR? I think this should be fixed by 1.13.0 for people relying on the old behavior (will likely break someone, even if it's just in tests suites)

**Comment by tianon** [comment]
@runcom that'd be reverting the entire Cobra conversion, which is likely going to cause a cascade of trouble :sweat_smile:

(I do have scripts that were broken by this, but you're probably right that I'm not the only one)

**Comment by runcom** [comment]
> @runcom that'd be reverting the entire Cobra conversion, which is likely going to cause a cascade of trouble

yeah, I thought there was just a PR which broke this. Can something be hacked/wired to fix this for the release?

**Comment by dnephin** [comment]
Ya, I'll continue to look for a fix.

**Comment by dnephin** [comment]
I've got a fix in #28443
